File: src/main/ruby/diffj/ast/method.rb
#!/usr/bin/jruby -w
# -*- ruby -*-

require 'java'
require 'diffj/ast/item'
require 'diffj/ast/function'

module DiffJ
  class MethodComparator < FunctionComparator

    METHOD_BLOCK_ADDED = "method block added"
    METHOD_BLOCK_REMOVED = "method block removed"

    VALID_MODIFIERS = [
                       ::Java::net.sourceforge.pmd.ast.JavaParserConstants::ABSTRACT,
                       ::Java::net.sourceforge.pmd.ast.JavaParserConstants::FINAL,
                       ::Java::net.sourceforge.pmd.ast.JavaParserConstants::NATIVE,
                       ::Java::net.sourceforge.pmd.ast.JavaParserConstants::STATIC,
                       ::Java::net.sourceforge.pmd.ast.JavaParserConstants::STRICTFP
                      ]
    
    def compare_modifiers from, to
      super from.parent, to.parent, VALID_MODIFIERS
    end

    def compare_parameters from, to
      super from.parameters, to.parameters
    end

    def compare_throws from, to
      super from, from.throws_list, to, to.throws_list
    end

    def get_block node
      node.find_child "net.sourceforge.pmd.ast.ASTBlock"
    end
    
    def compare_bodies from, to
      fromblock = get_block from
      toblock = get_block to

      if fromblock.nil?
        if toblock
          changed from, to, METHOD_BLOCK_ADDED
        end
      elsif toblock.nil?
        changed from, to, METHOD_BLOCK_REMOVED
      else
        compare_blocks from.fullname, fromblock, to.fullname, toblock
      end
    end

    def compare_blocks fromname, fromblock, toname, toblock
      fromcode = fromblock.get_child_tokens
      tocode = toblock.get_child_tokens

      compare_code fromname, fromcode, toname, tocode
    end

    def compare from, to
      compare_modifiers from, to
      compare_return_types from, to
      compare_parameters from, to

      compare_throws from, to
      compare_bodies from, to
    end
  end
end

File: src/main/java/org/incava/diffj/Method.java
package org.incava.diffj;

import java.util.List;
import net.sourceforge.pmd.ast.ASTBlock;
import net.sourceforge.pmd.ast.ASTFormalParameters;
import net.sourceforge.pmd.ast.ASTMethodDeclaration;
import net.sourceforge.pmd.ast.ASTMethodDeclarator;
import net.sourceforge.pmd.ast.ASTNameList;
import net.sourceforge.pmd.ast.SimpleNode;
import net.sourceforge.pmd.ast.Token;
import org.incava.diffj.params.Parameters;
import org.incava.pmdx.MethodUtil;
import org.incava.pmdx.ParameterUtil;
import org.incava.pmdx.SimpleNodeUtil;

public class Method extends Function implements Diffable<Method> {
    private final ASTMethodDeclaration method;
    private final ASTBlock block;

    public Method(ASTMethodDeclaration method) {
        super(method);
        this.method = method;
        this.block = (ASTBlock)SimpleNodeUtil.findChild(method, "net.sourceforge.pmd.ast.ASTBlock");
    }

    public void diff(Method toMethod, Differences differences) {
        compareAccess(toMethod, differences);
        compareModifiers(toMethod, differences);
        compareReturnTypes(toMethod, differences);
        compareParameters(toMethod, differences);
        compareThrows(toMethod, differences);
        compareBodies(toMethod, differences);
    }

    protected ASTFormalParameters getFormalParameters() {
        return MethodUtil.getParameters(method);
    }

    protected ASTNameList getThrowsList() {
        return MethodUtil.getThrowsList(method);
    }

    protected ASTBlock getBlock() {
        return block;
    }

    protected SimpleNode getReturnType() {
        return SimpleNodeUtil.findChild(method);
    }

    /**
     * This returns the full method name/signature, including parameters.
     */
    public String getName() {
        return MethodUtil.getFullName(method);
    }

    protected MethodModifiers getModifiers() {
        return new MethodModifiers(getParent());
    }

    protected void compareModifiers(Method toMethod, Differences differences) {
        MethodModifiers fromMods = getModifiers();
        MethodModifiers toMods = toMethod.getModifiers();
        fromMods.diff(toMods, differences);
    }

    protected boolean hasBlock() {
        return block != null;
    }

    protected void compareBodies(Method toMethod, Differences differences) {
        if (hasBlock()) {
            if (toMethod.hasBlock()) {
                compareCode(toMethod, differences);
            }
            else {
                differences.changed(method, toMethod.method, Messages.METHOD_BLOCK_REMOVED);
            }
        }
        else if (toMethod.hasBlock()) {
            differences.changed(method, toMethod.method, Messages.METHOD_BLOCK_ADDED);
        }
    }

    // protected void compareBlocks(String fromName, ASTBlock fromBlock, String toName, ASTBlock toBlock)
    // {
    //     tr.Ace.cyan("fromBlock", fromBlock);
    //     SimpleNodeUtil.dump(fromBlock, "");
    //     tr.Ace.cyan("toBlock", toBlock);
    //     SimpleNodeUtil.dump(toBlock, "");
    //     // walk through, looking for common if and for statements ...
    //     tr.Ace.cyan("aChildren(null)", SimpleNodeUtil.findChildren(fromBlock));
    //     tr.Ace.cyan("bChildren(null)", SimpleNodeUtil.findChildren(toBlock));        
    // }

    protected List<Token> getCodeTokens() {
        return SimpleNodeUtil.getChildTokens(block);
    }

    protected void compareReturnTypes(Method toMethod, Differences differences) {
        SimpleNode fromRetType    = getReturnType();
        SimpleNode toRetType      = toMethod.getReturnType();
        String     fromRetTypeStr = SimpleNodeUtil.toString(fromRetType);
        String     toRetTypeStr   = SimpleNodeUtil.toString(toRetType);

        if (!fromRetTypeStr.equals(toRetTypeStr)) {
            differences.changed(fromRetType, toRetType, Messages.RETURN_TYPE_CHANGED, fromRetTypeStr, toRetTypeStr);
        }
    }

    /**
     * This returns only the method name, without the parameters.
     */
    public String getMethodName() {
        ASTMethodDeclarator decl = MethodUtil.getDeclarator(method);
        return decl.getFirstToken().image;
    }

    public double getMatchScore(Method toMethod) {
        String fromName = getMethodName();
        String toName = toMethod.getMethodName();

        if (!fromName.equals(toName)) {
            return 0;
        }

        Parameters fromParams = getParameters();
        Parameters toParams = toMethod.getParameters();

        return fromParams.getMatchScore(toParams);
    }

    public String getAddedMessage() {
        return Messages.METHOD_ADDED;
    }

    public String getRemovedMessage() {
        return Messages.METHOD_REMOVED;
    }
}
